File: club/shmoke/client/cheats/personal/Velocity.java
package club.shmoke.client.cheats.personal;

import net.minecraft.network.play.server.S12PacketEntityVelocity;
import net.minecraft.network.play.server.S27PacketExplosion;
import club.shmoke.api.cheat.Cheat;
import club.shmoke.api.cheat.data.Property;
import club.shmoke.client.events.other.RecievePacketEvent;
import club.shmoke.client.events.render.HurtcamEvent;
import club.shmoke.api.event.interfaces.EventListener;

public class Velocity extends Cheat {
    public Property<Mode> mode = new Property(this, "Mode", Mode.NORMAL);
    public Property<Integer> hori = new Property<>(this, "Horizontal", 0, -200, 200, 1);
    public Property<Integer> vert = new Property<>(this, "Vertical", 0, 0, 200, 1);
    private Property<Boolean> hurtcam = new Property(this, "Hurtcam", true);
    public Velocity() {
        super("Velocity", Type.PERSONAL);
        this.description = "Negates velocity.";
    }

    double beforeHurtY = 0;

    @EventListener
    public void hurtcam(HurtcamEvent e) {
        if (!hurtcam.getValue())
            e.cancel();
    }

    @EventListener
    public void packet(RecievePacketEvent event) {
        if (mc.theWorld == null) return;
        switch (mode.getValue()) {

            case NORMAL:
                if (event.getPacket() instanceof S12PacketEntityVelocity) {
                    S12PacketEntityVelocity packet = (S12PacketEntityVelocity) event.getPacket();
                    double xz = hori.getValue() * 0.01;
                    double y = vert.getValue() * 0.01;
                    packet.x = (int) xz;
                    packet.y = (int) y;
                    packet.z = (int) xz;
                }

                if (event.getPacket() instanceof S27PacketExplosion) {
                    S27PacketExplosion packetExplosion = (S27PacketExplosion) event.getPacket();
                    packetExplosion.x = packetExplosion.y = packetExplosion.z = 0;
                }
                break;

            case AAC: {
                if (mc.thePlayer.hurtTime >= 1 && mc.thePlayer.hurtTime <= 8) { // 8
                    double yaw = mc.thePlayer.rotationYawHead;
                    yaw = Math.toRadians(yaw);
                    final double dX = -Math.sin(yaw) * 0.05;
                    final double dZ = Math.cos(yaw) * 0.05;
                    mc.thePlayer.motionX = dX;
                    mc.thePlayer.motionZ = dZ;
                }

                break;
            }

            case MMC: {
                if (mc.thePlayer.hurtTime > 0 && mc.thePlayer.hurtTime < 10) {
                    double yaw = mc.thePlayer.rotationYawHead;
                    yaw = Math.toRadians(yaw);
                    final double dX = -Math.sin(yaw) * 0.1;
                    final double dZ = Math.cos(yaw) * 0.1;
                    mc.thePlayer.setPosition(mc.thePlayer.posX + dX, mc.thePlayer.posY - (mc.thePlayer.hurtTime == 5 && !mc.thePlayer.onGround ? 0.05 : 0), mc.thePlayer.posZ + dZ);
                }
                break;
            }
            case HYPIXEL: {
                if (event.getPacket() instanceof S12PacketEntityVelocity || event.getPacket() instanceof S27PacketExplosion)
                    event.cancel();
                break;
            }
        }
    }

    public enum Mode {
        NORMAL, AAC, MMC, HYPIXEL
    }
}

File: fbiclient v2/club/shmoke/main/cheats/user/Velocity.kt
package club.shmoke.main.cheats.user

import club.shmoke.api.cheat.Category
import club.shmoke.api.cheat.Cheat
import club.shmoke.api.cheat.property.Property
import club.shmoke.api.event.EventHandler
import club.shmoke.main.events.PacketEvent
import club.shmoke.main.events.UpdateEvent
import net.minecraft.network.play.server.S12PacketEntityVelocity

/**
 * @author Christian
 */
class Velocity : Cheat("Velocity", 0, Category.USER, "Manipulates the knockback effect.") {

    private val mode = Property(this, "Mode", Mode.CANCEL)

    @EventHandler
    fun onUpdate(event: UpdateEvent) {
        suffix = mode.value.toString() + ""
        if (mode.value == Mode.PUSH) {
            if (mc.thePlayer.hurtTime in 1..4) {
                mc.thePlayer.motionX *= 0.8
                mc.thePlayer.motionZ *= 0.8
            } else if (mc.thePlayer.hurtTime == 6) {
                playerUtility.updatePosition(0.3, -0.2)
            }
        }
    }

    @EventHandler
    fun onPacket(event: PacketEvent) {
        if (mc.thePlayer == null || mc.theWorld == null) return
        if (event.type === PacketEvent.Type.RECEIVE) {
            if (event.packet is S12PacketEntityVelocity) {
                val packet = event.packet as S12PacketEntityVelocity?
                if (packet!!.entityID == mc.thePlayer.entityId) {
                    when (mode.value) {
                        Velocity.Mode.CANCEL -> event.cancel()
                        Velocity.Mode.GROUND -> {
                            if (mc.thePlayer.onGround)
                                playerUtility.updatePosition(0.0, -0.7)
                            else
                                playerUtility.updatePosition(0.0, -1.9)
                        }
                    }
                }
            }
        }
    }

    enum class Mode {
        CANCEL, GROUND, PUSH
    }

}
